crates/compiler/unify/src/fix.rs

```diff
@@ -0,0 +1,334 @@
+                            (Some((left_field, left_v)), Some((right_field, right_v))) => {
+                                assert_eq!(left_field, right_field, "fields do not unify");
+                                if let Ok(chain) =
+                                    help(subs, needle, left_v.into_inner(), right_v.into_inner())
+                                {
+                                    break Ok(chain);
+                                }
+                            }
+                            (None, None) => break Err(()),
+                            _ => internal_error!("fields differ; does not unify"),
+                        }
+                    }?;
+                    chain.push((left, right));
+                    Ok(chain)
+                }
+                (
+                    FunctionOrTagUnion(_left_tag_name, left_sym, left_var),
+                    FunctionOrTagUnion(_right_tag_name, right_sym, right_var),
+                ) => {
+                    assert_eq!(
+                        subs.get_subs_slice(*left_sym),
+                        subs.get_subs_slice(*right_sym)
+                    );
+                    let mut chain = help(subs, needle, *left_var, *right_var)?;
+                    chain.push((left, right));
+                    Ok(chain)
+                }
+                (TagUnion(left_tags, left_ext), TagUnion(right_tags, right_ext))
+                | (
+                    RecursiveTagUnion(_, left_tags, left_ext),
+                    RecursiveTagUnion(_, right_tags, right_ext),
+                )
+                | (TagUnion(left_tags, left_ext), RecursiveTagUnion(_, right_tags, right_ext))
+                | (RecursiveTagUnion(_, left_tags, left_ext), TagUnion(right_tags, right_ext)) => {
+                    let (left_it, _) = left_tags.sorted_iterator_and_ext(subs, *left_ext);
+                    let (right_it, _) = right_tags.sorted_iterator_and_ext(subs, *right_ext);
+                    assert_eq!(
+                        left_it.len(),
+                        right_it.len(),
+                        "tag lengths differ; does not unify"
+                    );
+
+                    for ((left_tag, left_args), (right_tag, right_args)) in left_it.zip(right_it) {
+                        assert_eq!(left_tag, right_tag);
+                        if let Ok(mut chain) =
+                            short_circuit(subs, needle, left_args.iter(), right_args.iter())
+                        {
+                            chain.push((left, right));
+                            return Ok(chain);
+                        }
+                    }
+
+                    Err(())
+                }
+                (EmptyRecord, EmptyRecord)
+                | (EmptyTagUnion, EmptyTagUnion) => Err(()),
+                _ => internal_error!(
+                    "structures {:?} and {:?} do not unify; they should never have been involved in fixing!",
```

Comment:
is that true? e.g. `EmptyRecord` can unify with a record with zero fields. `FunctionOrTagUnion` also has more unification options than what this pattern match does.

Reply:
You're right, it's not true